hngprojects/hng_boilerplate_python_fastapi_web
# Feat : register super admin User

> > ## Description  : This update adds a new endpoint POST that registers super admin users
> > ## How Has This Been Tested?
> > 
> > 1. **Unit Tests**:
> >    
> >    * Implemented unit tests for the register super admin user endpoint
> >    * I made sure it validates data received from the user properly
> >    * I made it gave the appropriate response on Success
> > 
> > ## Types of Changes
> > 
> > * [ ]   Bug fix (non-breaking change which fixes an issue)
> > * [x]   New feature (non-breaking change which adds functionality)
> > * [ ]   Breaking change (fix or feature that would cause existing functionality to change)
> > 
> > ## Checklist
> > 
> > * [x]   My code follows the code style of this project.
> > * [x]   My change requires a change to the documentation.
> > * [x]   I have updated the documentation accordingly.
> > * [x]   I have read the CONTRIBUTING document.
> > * [x]   I have added tests to cover my changes.
> > * [x]   All new and existing tests passed.
> > 
> > ## How to Run and Test the Project
> > ### Prerequisites
> > Ensure you have the following installed on your machine:
> > 
> > * Python 3.10
> > * PostgreSQL
> > * Git
> > * Virtual env (optional but recommended)
> > 
> > ### Setup Instructions
> > 
> > 1. **Clone the Repository**
> >    ```shell
> >    git clone <repository_url>
> >    cd <repository_directory>
> >    ```
> > 2. **Create and Activate a Virtual Environment (optional but recommended)**
> >    ```shell
> >    python -m venv .venv
> >    source .venv/bin/activate  # On Windows use `.venv\Scripts\activate`
> >    ```
> > 3. **Install Dependencies**
> >    ```shell
> >    pip install -r requirements.txt 
> >    ```
> > 4. **Configure Environment Variables**
> >    Create a `.env` file in the root directory and add the following environment variables:
> > 
> > PYTHON_ENV=dev
> 
> > DB_TYPE=postgresql
> > DB_NAME=postgres
> > DB_USER=postgres
> > DB_PASSWORD=Winner174
> > DB_HOST="localhost"
> > DB_PORT=5432
> > MYSQL_DRIVER=
> > DB_URL=postgresql://postgres:Winner174@localhost:5432/postgres
> > SECRET_KEY = ""
> > ALGORITHM = HS256
> > ACCESS_TOKEN_EXPIRE_MINUTES = 10
> > JWT_REFRESH_EXPIRY=5
> > APP_URL=
> 
> > MAIL_USERNAME=""
> > MAIL_PASSWORD=""
> > MAIL_FROM=""
> > MAIL_PORT=465
> > MAIL_SERVER="smtp.gmail.com"
> > ```
> > 5. **Run Database Migrations**
> > ```shell
> > alembic upgrade head
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > 6. **Start the FastAPI Server**
> >    ```shell
> >    python main.py
> >    ```
> >    
> >    
> >        
> >          
> >        
> >    
> >          
> >        
> >    
> >        
> >      
> >    The server will start at `http://127.0.0.1:7001`.
> > 
> > ### Testing Instructions
> > 
> > 1. **Run Unit Tests with Pytest**
> >    ```shell
> >    pytest
> >    ```
> >    
> >    
> >        
> >          
> >        
> >    
> >          
> >        
> >    
> >        
> >      
> >    This will run all the test cases and provide a report of the test results.
> 
> > 2. **Test the POST superadmin endpoint**
> >    
> >    * Open Postman and create a new GET request to `http://127.0.0.1:7001/api/v1/superadmin/register`.
> >    * Send the request.
> >    * You should receive a response indicating  that a user has been created and the details of the created user:
> >      ```json
> >      {
> >         {
> >      ```
> 
> ```
>          "status_code": 201,
>            "success": true,
>           "message": "User Created Successfully",
>           "data": {
>                 "id": "066a175e-165d-7908-8000-4dcba2606466",
>                 "first_name": "Utibe",
>                  "last_name": "Uboh",
>                   "username": "Marvelous16",
>                   "email": "utibesolomon17@gmail.com",
>                      "created_at": "2024-07-24 22:45:05.093033+01:00"
> ```
> 
> } }
> 
> > ```
> >          }
> >      ]
> >  }
> >  ```
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > ### Notes
> > 
> > * Ensure your PostgreSQL database is running and accessible with the credentials provided in the `.env` file.
> > * If you encounter any issues, please check the logs for detailed error messages and verify that all dependencies are correctly installed.
